### PR TITLE
[schemadb] put max_open_files into storage-config

### PR DESCRIPTION
**Pull request by lightmark**
The number of files keeps increasing so that the RSS increases unlimitedly due to index/filter_block per file.

I'll later refactor it.

## Motivation

https://github.com/libra/libra/issues/6644

### CODE REVIEW

**Comment by mgorven** [comment]
How complex is it to make this configurable? I'm just thinking there might be unintended consequences (reduced read perf?) and we might want to tune this in config.

**Comment by lightmark** [comment]
It is easy. But I guess in the future we may want a option files exclusively for rocksdb and here `max_open_files` is just one item.

**Comment by davidiw** [comment]
> @mgorven It is easy. But I guess in the future we may want an option file exclusively for rocksdb and `max_open_files` is just one item in several options.

I think it is okay to create a struct in the storage config call rocks config that way you don't have to worry about this.

**Comment by msmouse** [comment]
The memory consumption 1% of the DB size on disk, which is well affordable and most probably worth it if people actually query the full history on fullnodes. And I guess we are not too sure in or two years we will actually hit a larger DB in production.

I have no problem making it an option that's tunable (okay, to be honest I'd avoid if you ask me), but I'd make our default value a pretty large number (5k or 10k).

@mgorven  How much memory do you think we can afford for this purpose?

**Comment by mgorven** [comment]
> I have no problem making it an option that's tunable (okay, to be honest I'd avoid if you ask me), but I'd make our default value a pretty large number (5k or 10k).
> 
> @mgorven How much memory do you think we can afford for this purpose?

I agree that we can go larger than 1k (based on the current test fullnode which hasn't exceeded 1G). The lowest memory allocation currently is 5.25G (on Azure). We need some space for the OS page cache and transaction handling under load, so somewhere around 3G should be fine.

Given the large impact this has on memory usage I think this needs to be configurable. For public fullnodes users may want to reduce memory usage, and if they're only querying recent data setting a low limit shouldn't affect read performance.

**Comment by mgorven** [comment]
I think 10k is a good default.

**Comment by lightmark** [comment]
Then let's add a struct in the option.

File: config/src/config/storage_config.rs [hunk 1]
```diff
@@ -8,6 +8,33 @@ use std::{
     path::PathBuf,
 };
 
+#[derive(Copy, Clone, Debug, Deserialize, PartialEq, Serialize)]
+#[serde(default, deny_unknown_fields)]
+pub struct RocksdbConfig {
+    /// Port selected RocksDB options for tuning underlying rocksdb instance.
+    /// see https://github.com/facebook/rocksdb/blob/master/include/rocksdb/options.h
+    /// for detailed explanations.
```
Comment (msmouse) [inline config/src/config/storage_config.rs:16]: should be before the whole struct?

File: config/src/config/storage_config.rs [hunk 2]
```diff
@@ -8,6 +8,33 @@ use std::{
...
+    pub create_if_missing: bool,
+    pub create_if_missing_column_families: bool,
```
Comment (msmouse) [inline config/src/config/storage_config.rs:18]: `open()`, `open_readonly()` and `open_as_secondary()` use different values for these options, I don't think we should pass them in?

File: config/src/config/storage_config.rs [hunk 3]
```diff
@@ -8,6 +8,33 @@ use std::{
...
+    pub max_open_files: i32,
+    pub max_total_wal_size: u64,
+}
+
+impl Default for RocksdbConfig {
+    fn default() -> Self {
+        Self {
+            create_if_missing: true,
+            create_if_missing_column_families: true,
```
Comment (mgorven) [inline config/src/config/storage_config.rs:27]: When would these ever be changed?

File: config/src/config/storage_config.rs [hunk 4]
```diff
@@ -8,6 +8,33 @@ use std::{
...
+#[derive(Copy, Clone, Debug, Deserialize, PartialEq, Serialize)]
+#[serde(default, deny_unknown_fields)]
+pub struct RocksdbConfig {
+    /// Port selected RocksDB options for tuning underlying rocksdb instance.
+    /// see https://github.com/facebook/rocksdb/blob/master/include/rocksdb/options.h
+    /// for detailed explanations.
+    pub create_if_missing: bool,
+    pub create_if_missing_column_families: bool,
+    pub max_open_files: i32,
+    pub max_total_wal_size: u64,
```
Comment (davidiw) [inline config/src/config/storage_config.rs:20]: what's a wal?

File: consensus/src/consensusdb/mod.rs [hunk 5]
```diff
@@ -36,7 +36,10 @@ impl ConsensusDB {
 
         let path = db_root_path.as_ref().join("consensusdb");
         let instant = Instant::now();
-        let db = DB::open(path.clone(), "consensus", column_families)
+        let mut opts = Options::default();
+        opts.create_if_missing(true);
+        opts.create_missing_column_families(true);
```
Comment (davidiw) [inline consensus/src/consensusdb/mod.rs:41]: these default as true, why are we modifying it?

File: storage/libradb/src/lib.rs [hunk 6]
```diff
@@ -93,6 +94,15 @@ fn error_if_too_many_requested(num_requested: u64, max_allowed: u64) -> Result<(
...
 
+fn gen_rocksdb_options(config: &RocksdbConfig) -> Options {
+    let mut db_opts = Options::default();
+    db_opts.create_if_missing(config.create_if_missing);
+    db_opts.create_missing_column_families(config.create_if_missing_column_families);
+    db_opts.set_max_open_files(config.max_open_files);
+    db_opts.set_max_total_wal_size(config.max_total_wal_size);
+    db_opts
+}
+
```
Comment (davidiw) [inline storage/libradb/src/lib.rs:105]: can we use an std::convert::Into instead?

File: config/src/config/storage_config.rs (hunk 4 above)
Comment (lightmark) [reply config/src/config/storage_config.rs:20]: write_ahead_log

File: config/src/config/storage_config.rs (hunk 3 above)
Comment (lightmark) [reply config/src/config/storage_config.rs:27]: for val/fullnode, they don't. But for my tool to access db, or open_read_only(), I don't want to create a db if it doesn't exist there. It should be an error "DB not found"

File: consensus/src/consensusdb/mod.rs (hunk 5 above)
Comment (lightmark) [reply consensus/src/consensusdb/mod.rs:41]: default is false. Here ConsensusDB take raw schemaDB so it doesn't do the conversion.

File: storage/libradb/src/lib.rs (hunk 6 above)
Comment (lightmark) [reply storage/libradb/src/lib.rs:105]: I don't think so. Please tell me where it should be. I weighed different solutions. 
the `Into` trait must be implemented in either crate, rocksdb binding (which is third party) or libra-config (do you want to incorporate rocksdb dependency into `libra-config`?)

**Review by msmouse — APPROVED** [review]

**Pull request closed without merging**